### ISSUE TITLE
exception and thread safety

### ISSUE THREAD

**Issue by casperdcl** [tqdm/tqdm#548]
- make `Exception` safe
    + wrapping iterables that are expected to raise Exceptions #469
    + probably also fixes immediate delete case #406 (formerly `wont-fix`)
- enforce and share locks across threads etc to suppress warnings #481

**Comment by acbart** [comment]
Yeah, interesting behavior. It seems like if an exception gets triggered at any point, inner progressbars get spawned on new lines instead of staying on the previous line.

**Comment by casperdcl** [comment]
@fabianp you mean this doesn't work for you?

```python
try:
    with tqdm(...) as t:
        for i in t:
            ...
except KeyboardInterrupt:
    t.close()
    raise
t.close()
```

**Comment by aldanor** [comment]
Any further thoughts on this?

I've copied and pasted this hundreds of times now... this gets even worse if you actually need direct access to progressbar object, then you have to use contextmanagers etc. Feels like something of the sort should really be hard-baked into tqdm itself?

```python
def tqdm_iter(iterable):
    pbar = tqdm.tqdm(iterable)
    try:
        yield from pbar
    finally:
        pbar.close()
```

**Comment by xoxoxo13102020** [comment]
`from tqdm import tqdm as tqdm`

worked for me

**Comment by casperdcl** [comment]
It should be exception safe (`tqdm>=4.45.0` or even better `tqdm>=4.50.0`), it's just built-in thread safety which sometimes needs to be manually dealt with.
